net/src/ipv4/prefix.rs

```diff
@@ -0,0 +1,665 @@
+
+/// An error indicating that an invalid prefix length was provided.
+#[derive(Debug, thiserror::Error)]
+#[repr(transparent)]
+pub enum InvalidIpv4PrefixLength {
+    /// The provided prefix is too long to form a legal [`Ipv4PrefixLen`]
+    #[error("invalid prefix length {0} is invalid, max is {MAX}", MAX = Ipv4PrefixLen::MAX_LEN)]
+    TooLong(u8),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidIpv4Network {
+    /// The provided network description contains set non-network bits
+    #[error("Address {0}/{1} contains non network bits")]
+    AddressContainsNonNetworkBits(Ipv4Addr, Ipv4PrefixLen),
+    /// The provided prefix length is invalid
+    #[error(transparent)]
+    InvalidPrefix(InvalidIpv4PrefixLength),
+}
+
+/// An error indicating that an invalid network was provided.
+#[derive(Debug, thiserror::Error)]
+pub enum Ipv4PrefixParseError {
+    /// failure to interpret string as an ip and a prefix length
+    #[error(transparent)]
+    AddrParseError(AddrParseError),
+    /// invalid ip or prefix length
+    #[error(transparent)]
+    InvalidIpv4Network(InvalidIpv4Network),
+}
+
+impl Ipv4PrefixLen {
+    /// The largest possible prefix length for IPv4 (i.e. /32)
+    pub const MAX_LEN: u8 = 32;
+    /// The largest possible prefix length for IPv4 (i.e. /32)
+    pub const MAX: Self = Self(Self::MAX_LEN);
+    /// The minimum possible prefix length for IPv4 (i.e. /0)
+    pub const MIN: Self = Self(0);
+
+    /// Constructor which asserts if the provided length is invalid.
+    /// Useful in const contexts where you are sure you won't panic.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the provided length is greater than [`Ipv4PrefixLen::MAX_LEN`].
+    #[must_use]
+    pub const fn new_assert(len: u8) -> Self {
+        assert!(len <= Self::MAX_LEN, "invalid prefix length");
+        Self(len)
+    }
+
+    /// Constructor which checks that the provided length is valid.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`InvalidIpv4PrefixLength::TooLong`] if the provided length is greater than
+    /// [`Ipv4PrefixLen::MAX_LEN`].
+    pub const fn try_new(len: u8) -> Result<Ipv4PrefixLen, InvalidIpv4PrefixLength> {
+        if len > Self::MAX_LEN {
+            return Err(InvalidIpv4PrefixLength::TooLong(len));
+        }
+        Ok(Ipv4PrefixLen(len))
+    }
+
+    /// Interpret the [`Ipv4PrefixLen`] as a `u8`
+    #[must_use]
+    pub const fn as_u8(&self) -> u8 {
+        self.0
+    }
+}
+
+impl TryFrom<u8> for Ipv4PrefixLen {
+    type Error = InvalidIpv4PrefixLength;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        Ipv4PrefixLen::try_new(value)
+    }
+}
+
+impl From<Ipv4PrefixLen> for u8 {
+    fn from(value: Ipv4PrefixLen) -> Self {
+        value.0
+    }
+}
+
+impl From<Ipv4Addr> for Ipv4Prefix {
+    fn from(value: Ipv4Addr) -> Self {
+        Ipv4Prefix::new_assert(value.octets(), Ipv4PrefixLen::MAX.as_u8())
```

Comment:
I see this might cause trouble.

Reply:
Agreed, Daniel and I discussed not using this in the code.  I'd be in favor of dropping it, but I am not hell bent on that.  But it certainly should not be used on the datapath.